mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,499 @@
+use core::panic;
+use serde_json::Value;
+
+use clap::{Arg, Command};
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+
+#[derive(Debug, PartialEq)]
+pub enum Filter {
+    Identity,
+    ObjectIdentifier(String),
+    ArrayIndex(isize),
+    ArraySlice(Option<isize>, Option<isize>),
+    ArrayIterator,
+    Add,
+    Length,
+    Del(Box<Filter>),
+    // Other filters as needed
+}
+
+//name:
+//obj_identify -- .<KEY_NAME>
+//array_index --  .[i32]
+//array_slice -- .[<start_number>:<end_number>]
+//pipe -- pass output of one filter to another
+// -- handle in main loop?
+//array_iterator -- turn an array into an iterator
+// -- effectively expands an array into multiple objects
+
+//add -- adds all values in an array
+//length -- returns the length of the input
+//del -- del(.[<indexes>]) -- return everything but deleted
+
+pub fn apply_filters(values: Vec<Value>, filters: &[Filter]) -> Result<Vec<Value>, String> {
+    let mut current_values = values;
+
+    for filter in filters {
+        let mut next_values = Vec::new();
+        for value in current_values {
+            let result = apply_single_filter(value, filter)?;
+            next_values.extend(result);
+        }
+        current_values = next_values;
+    }
+
+    Ok(current_values)
+}
+
+pub fn apply_single_filter(value: Value, filter: &Filter) -> Result<Vec<Value>, String> {
+    match filter {
+        Filter::Identity => Ok(vec![value]),
+        Filter::ObjectIdentifier(key) => {
+            if let Value::Object(map) = value {
+                match map.get(key) {
+                    Some(v) => Ok(vec![v.clone()]),
+                    None => Err(format!("Key '{}' not found in object.", key)),
+                }
+            } else {
+                Err("Value is not an object.".to_string())
+            }
+        }
+        Filter::ArrayIndex(index) => {
+            if let Value::Array(array) = value {
+                let idx = if *index < 0 {
+                    array.len() as isize + index
+                } else {
+                    *index
+                };
+                if idx >= 0 && (idx as usize) < array.len() {
+                    Ok(vec![array[idx as usize].clone()])
+                } else {
+                    Err("Array index out of bounds.".to_string())
+                }
+            } else {
+                Err("Value is not an array.".to_string())
+            }
+        }
+        Filter::ArraySlice(start_opt, end_opt) => {
+            if let Value::Array(array) = value {
+                let len = array.len() as isize;
+                let start = start_opt.unwrap_or(0);
+                let end = end_opt.unwrap_or(len);
+                let start = if start < 0 { len + start } else { start };
+                let end = if end < 0 { len + end } else { end };
```

Comment:
took me a hot sec, but I see what you're doing here - I almost feel like it would have been better if you just inlined each of these for each unique variable